lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/ReduceOpsRewritePattern.cpp

```diff
@@ -0,0 +1,62 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Dialect/TTNN/Transforms/Workarounds/Decomposition/ReduceOpsRewritePattern.h"
+
+#include <algorithm>
+#include <unordered_set>
+
+namespace mlir::tt::ttnn::workarounds::decomposition {
+
+std::vector<int64_t>
+getReduceDims(const std::optional<mlir::ArrayAttr> &dimArg) {
+  std::vector<int64_t> reduceDims;
+  if (!dimArg.has_value()) {
+    return reduceDims;
+  }
+
+  for (const mlir::Attribute &reduceDim : dimArg.value()) {
+    reduceDims.push_back(mlir::cast<mlir::IntegerAttr>(reduceDim).getInt());
+  }
+
+  return reduceDims;
+}
+
+std::vector<int64_t>
+calculateNewReduceShape(const std::optional<mlir::ArrayAttr> &dimArg,
+                        const RankedTensorType &inputType) {
+  std::vector<int64_t> outputShapeVec = inputType.getShape().vec();
+  std::vector<int64_t> reduceDims = getReduceDims(dimArg);
+
+  if (reduceDims.empty()) {
+    std::fill(outputShapeVec.begin(), outputShapeVec.end(), 1);
+  } else {
+    for (const int64_t reduceDim : reduceDims) {
+      outputShapeVec[reduceDim < 0 ? outputShapeVec.size() -
+                                         static_cast<size_t>(-reduceDim)
+                                   : static_cast<size_t>(reduceDim)] = 1;
```

Comment:
Vectors are indexed by `size_t` which can change depending on the system, that is why I chose to cast into `size_t` instead of `int64_t`. It is not unsafe because it is what will happen in the background anyways. Also, I can't imagine tensor ever having so many dimensions that this could be a problem :) I've decomposed the calculation into something easier to understand, let me know if you still see issues.